### ISSUE TITLE
`@apollo/client` should be a peer dependency

### ISSUE THREAD

**Issue by andreialecu** [jaydenseric/apollo-upload-client#251]
It doesn't seem to be correct that `@apollo/client` is a regular dependency.

https://github.com/jaydenseric/apollo-upload-client/blob/6dec8edc2bb1c97514e1577b84530922a40608bc/src/public/createUploadLink.js#L3-L12

Instead, it should be in `peerDependencies`.

I noticed this while exploring an app's dependencies, and I noticed that `@apollo/client` was included twice in the resulting installation.

**Comment by jaydenseric** [comment]
I'm open minded to make it a peer dependency, but it's surprising you’re having problems with the way it is now as a dependency.

If you run `npm ls @apollo/client`, what is the result? There should be a single version, deduped everywhere it’s a dependency.

If you try to pin your apps `@apollo/client` dependency to a version older than specified here (`^3.2.5`) you will get multiple versions installed. In that case, you should update as the current version is v3.3.13.

**Comment by andreialecu** [comment]
@jaydenseric as an example, if an user installs `apollo-upload-client` today, it will bring `@apollo/client` with it and pin it in to that specific version in the lockfile.

Now later on if the user bumps `@apollo/client` to a higher version, the one used by `apollo-upload-client` will not update one, because it was pinned previously. The only way to get it to update would be to clear the lockfile or manually update it.  I'm not sure if `npm` handles it by default, but `yarn` doesn't. Also it would be relatively unsafe and defeats the prupose of a lockfile. 

Yarn 2 has a `yarn dedupe` command for this purpose, for example.

This creates package duplication over time. Additionally, I think it's best to not load/parse `@apollo/client` twice, since the host project will already have it installed anyway.

**Comment by andreialecu** [comment]
Here's an example using yarn:

```sh
cd `mktemp -d`
yarn init
yarn add @apollo/client
yarn add apollo-upload-client
# notice downgrade below
yarn add @apollo/client@3.3.12 

yarn why @apollo/client
```

Output is:

```
yarn why v1.22.10
=> Found "@apollo/client@3.3.12"
info Has been hoisted to "@apollo/client"
info This module exists because it's specified in "dependencies".
info Disk size without dependencies: "5.16MB"
info Disk size with unique dependencies: "6.47MB"
info Disk size with transitive dependencies: "6.7MB"
info Number of shared dependencies: 14
=> Found "apollo-upload-client#@apollo/client@3.3.13"
info This module exists because "apollo-upload-client" depends on it.
info Disk size without dependencies: "5.16MB"
info Disk size with unique dependencies: "6.47MB"
info Disk size with transitive dependencies: "6.7MB"
info Number of shared dependencies: 14
✨  Done in 0.09s.
```

**Comment by jaydenseric** [comment]
@kevinatown yes, it's on the radar; I think about this issue and PR every day. Originally I had to finish up publishing [`graphql-react` v13](https://github.com/jaydenseric/graphql-react/releases/tag/v13.0.0) and updating related projects in it's mini ecosystem.

Now that is mostly done, I have prioritized publishing new major versions of all my packages, starting with dev tools (so they can be updated dev dependencies in all the other packages) dropping Node.js v10 support now that it's EOL today. This is kind of a big deal, because it involves migrating to 100% pure ESM (at least for dev tools; some of the prod packages might remain dual mode for a little bit due to Next.js bugs with pure ESM). This is something myself and a lot of other package maintainers have committed to doing together to rip the bandaid off so it's somewhat time sensitive:

- https://github.com/sindresorhus/meta/discussions/15#discussioncomment-358440
- https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#pure-esm-package
- https://blog.sindresorhus.com/get-ready-for-esm-aa53530b3f77

As I'm publishing major new versions in bulk for 20-ish packages, this is a good opportunity to update a bunch of things to do with test scripts, CI config, JSDoc API doc generation, etc. at the same time to get them all consistent and avoid multiple passes of work and maybe major releases across the whole fleet. I've been perfecting my core dev dependency packages, and will soon be able to work up to updating production packages like this one soon.

I have been working full time on open source for nearly 6 months straight now, so rest assured I have no distractions and am getting to this PR as fast as I can :)
